include/aws/common/math.h

```diff
@@ -0,0 +1,189 @@
+        return ~(uint64_t)0;
+    } else {
+        return x;
+    }
+#endif
+}
+
+/**
+ * Multiplies a * b and returns the truncated result in *r. If the result
+ * overflows, returns 0, else returns 1.
+ */
+static inline int aws_common_mul_u64_checked(uint64_t a, uint64_t b, uint64_t *r) {
+#if defined(__x86_64__) && defined(__GNUC__)
+    // We can use inline assembly to do this efficiently on x86-64 and x86.
+
+    int flag;
+    uint64_t result;
+    __asm__(
+        "mulq %q3\n" // rax * b, result is in RDX:RAX, OF=CF=(RDX != 0)
+        "setno %b1\n" // rdx = (OF = 0)
+        "and $0xFF, %1\n" // mask out flag
+        : /* out: %rax */ "=a" (result), "=d" (flag)
+        : /* in: reg or memory (possibly rax/rdx), imm/reg/memory for 2nd operand */
+          "a" (a), "g" (b)
+        : /* clobbers: cc */ "cc"
+    );
+    *r = result;
+    return flag;
+#else
+    uint64_t x = a * b;
+    *r = x;
+    if (a != 0 && (a > 0xFFFFFFFF || b > 0xFFFFFFFF) && x / a != b) {
+        return 0;
+    } else {
+        return 1;
+    }
+#endif
+}
+
+/**
+ * Multiplies a * b. If the result overflows, returns 2^32 - 1.
+ */
+static inline uint32_t aws_common_mul_u32_saturating(uint32_t a, uint32_t b) {
+#if (defined(__i386__) || defined(__x86_64__)) && defined(__GNUC__)
+    // We can use inline assembly to do this efficiently on x86-64 and x86.
+
+    // we specify edx as an output, rather than a clobber, because we want to allow it
+    // to be allocated as an input register
+    uint32_t edx;
+    __asm__(
+        "mull %k3\n" // eax * b, result is in EDX:EAX, OF=CF=(EDX != 0)
+#ifdef __x86_64__
+        "cmovc %k4, %%eax\n"
+#else
+        // cmov isn't guaranteed to be available on x86-32
+        "jno .1f\n"
+        "mov $0xFFFFFFFF, %%eax\n"
+        ".1f:"
+#endif
+        : /* out: %rax = result/a, out: rdx (ignored) */ "=a" (a), "=d" (edx)
+        : /* in: operand 1 in eax */ "a" (a),
+          /* in: operand 2 (imm/reg/memory) */ "g" (b)
+#ifdef __x86_64__
+        // Saturation value for CMOV to use
+            , "r" (0xFFFFFFFF)
+#endif
+        : /* clobbers: cc */ "cc"
+    );
+    (void)edx; // suppress unused warnings
+    return a;
+#else
+    uint32_t x = a * b;
+    if (a != 0 && (a > 0xFFFF || b > 0xFFFF) && x / a != b) {
+        return ~(uint32_t)0;
+    } else {
+        return x;
+    }
+#endif
+}
+
+/**
+ * Multiplies a * b and returns the result in *r. If the result overflows,
+ * returns 0, else returns 1.
+ */
+static inline int aws_common_mul_u32_checked(uint32_t a, uint32_t b, uint32_t *r) {
+#if (defined(__i386__) || defined(__x86_64__)) && defined(__GNUC__)
+    // We can use inline assembly to do this efficiently on x86-64 and x86.
+    uint32_t result;
+    int flag;
+    // Note: We use SETNO which only takes a byte register. To make this easy,
+    // we'll write it to dl (which we throw away anyway) and mask off the high
+    // bits.
+    __asm__(
```

Comment:
Since these are in a static inline function (and thus could be used inside an inner loop) should they by volatile as well?

Reply:
I should have said, will be substituted as is

Reply:
I don't understand, why is volatile necessary in any of this?

Reply:
Too keep it from getting moved by the compiler if it’s inline in certain scenarios we haven’t predicted?

Reply:
If you're referring to `asm volatile`, this is only needed if the code has desired side effects other than its out parameters. Here, though, we're only interested in those out parameters; we're perfectly happy to have the compiler do things like delete it as part of dead value elimination, so we should avoid `volatile`.